webdav.go

```diff
@@ -78,6 +105,45 @@ func (wd *WebDAV) Provision(ctx caddy.Context) error {
 	return nil
 }
 
+// os.FileInfo wrapper to provide webdav.ETager implementation.
+type WebDAVFileInfo struct{ FileInfo }
+
+func (d WebDAVFileInfo) ETag(ctx context.Context) (string, error) {
+	return calculateEtag(d.FileInfo), nil
+}
+
+// webdav.File wrapper to make it able to return WebDAVFileInfo other than plain os.FileInfo.
+type WebDAVFileWrapper struct{ WebDAVFile }
+
+func (fw WebDAVFileWrapper) Stat() (os.FileInfo, error) {
+	fi, err := fw.WebDAVFile.Stat()
+	if err != nil {
+		return fi, err
+	}
+	return WebDAVFileInfo{FileInfo: fi}, nil
+}
+
+// webdav.Dir wrapper to return properly wrapped structs to override the ETag calculation.
+type WebDAVDirWrapper struct {
+	WebDAVDir
+}
+
+func (d WebDAVDirWrapper) OpenFile(ctx context.Context, name string, flag int, perm os.FileMode) (webdav.File, error) {
+	f, err := d.WebDAVDir.OpenFile(ctx, name, flag, perm)
+	if err != nil {
+		return f, err
+	}
+	return WebDAVFileWrapper{WebDAVFile: f}, err
```

Comment:
Could this end up hiding that `f` is a `os.File` so that optimizations like sendfile can't be used?

Reply:
That's true. Experiments reveal a huge performance impact. Unfortunately I don't know how to improve the situation.

Reply:
@wader Some, but not super directly. Check out these links:

- https://github.com/caddyserver/caddy/issues/4731
- https://github.com/caddyserver/caddy/pull/6565/files

I'm trying to conclude whether `*os.File` is strictly required, or if you just need to implement ReaderFrom/WriterTo interfaces.

Reply:
I tried some implementations but non of them works. Looks like it has to be either `*os.File` or `*io.LimitedReader`. Sendfile is only utilized in TCPConn's side. My hypothesis: some of the code are calling ReadFrom directly instead of calling io.Copy, in which case the wrapper's implementation(WriterTo) has a lower priority and only receives masked structs without ReadFrom.
<https://cs.opensource.google/go/go/+/refs/tags/go1.23.4:src/io/io.go;l=383-388>
<https://cs.opensource.google/go/go/+/master:src/net/sendfile_linux.go;l=25-35;bpv=0;bpt=1>
<https://cs.opensource.google/go/go/+/master:src/net/tcpsock_posix.go;l=47-55>
<https://cs.opensource.google/go/go/+/master:src/net/net.go;l=765-767>

Just found that it seems a simpler solution to add a path matcher `*/` restricting file_server to only serve listings. In that way, all file etags are generated by webdav server.

Reply:
@hyx0329 you should wrap the response writer as well, and define a `ReadFrom` on it.

You can refer to this as a reference:

```
type davWriter struct {
	http.ResponseWriter
}

func (w *davWriter) ReadFrom(r io.Reader) (int64, error) {
	// check ReaderFrom
	rr, _ := w.ResponseWriter.(io.ReaderFrom)
	// used by http.ServeContent
	lr, ok := r.(*io.LimitedReader)
	if ok {
		// promote the embedded *os.File
		dw, ok := lr.R.(*WebDAVFileWrapper)
		if df != nil {
			lr.R = df.WebDAVFile
		}
	}
	if rr != nil {
		return rr.ReadFrom(r)
	}
	return io.Copy(w.ResponseWriter, r)
}
```

Reply:
Matching directories to calculate the etag is wrong. As webdav will calculate the etag when serving the contents of the file. The etag here will differ from the etag from the file listing.

You can serve the file using the file server. But that requires explicit configuration and is not for everyone.

Reply:
Thanks! Now strace shows sendfile is utilized.

Reply:
Wow, nice! Thanks @WeidiDeng !